parse.php

```diff
@@ -1,97 +1,149 @@
+$heading = parseToArray($xpath, 'title', 'h1');
 
-// libxml_use_internal_errors(true); // Yeah if you are so worried about using @ with warnings
+// extract and save the document meta tags
+$tags = get_meta_tags($url);
 
+// extract document meta properties
 $query = '//*/meta[starts-with(@property, \'og:\')]';
 $metas = $xpath->query($query);
 foreach ($metas as $meta) {
     $property = $meta->getAttribute('property');
     $content = $meta->getAttribute('content');
     $rmetas[$property] = $content;
 }
-//var_dump($rmetas);
 
-$domain    = explode("/", $url,4);
-//echo "::: domain: ". $domain[2] . "\n";
+// decompose URL into its constituent parts
+$domain = explode("/", $url, 4);
 
-$domain[2] = strtoupper(preg_replace("/^www\./","", $domain[2]));
+// strip leading 'www' from the URL as necessary
+$domain[2] = preg_replace("/^www\./", "", $domain[2]);
 
-//echo "::: domain: ". $domain[2] . "\n";
-//echo "\n\n";
+// prepend the site name to the title meta property and make that the title of the (cached) URL seen by FB
+$rmetas["og:title"] = $domain[2] ." -- ". $rmetas["og:title"];
+$heading[1] = $rmetas["og:title"];
 
-$rmetas["og:title"] = $domain[2] ." // ". $rmetas["og:title"];
+// attempt to maintain the document language
+$lang = preg_replace("/_/", "-", $rmetas['og:locale']);
 
-$heading[1] = $rmetas["og:title"];
+// an array key with ':' in the name will cause the PHP heredoc parser to choke
+$tags = replaceArrayKeys($tags);
+$rmetas = replaceArrayKeys($rmetas);
 
-echo "<!DOCTYPE html>\n <html lang='en'>\n  <head>\n\n";
+// output the (cached) URL metadata
+echo <<<"EOF"
+<!DOCTYPE html>
+<html lang="$lang">
+<head>
 
-echo "<!-- ". $url . "-->\n\n";
+<!-- $url -->
 
-echo "   <title>". $heading[1] ."</title>\n";
-echo '   <meta data-rh="true" name="description" content="'   . $tags["description"].   '">' ."\n";
-echo '   <meta data-rh="true" name="twitter:image" content="' . $tags["twitter:image"]. '">' ."\n";
+<title>$heading[1]</title>
 
-echo '   <meta data-hr="true" property="og:title"       content="' . $rmetas["og:title"]       . '">' ."\n";
-echo '   <meta data-hr="true" property="og:description" content="' . $rmetas["og:description"] . '">' ."\n";
-echo '   <meta data-hr="true" property="og:image"       content="' . $rmetas["og:image"]       . '">' ."\n";
+<meta data-rh="true" name="description" content="$tags[description]">
+<meta data-rh="true" name="twitter:image" content="$tags[twitter_image]">
+<meta data-hr="true" property="og:title" content="$rmetas[og_title]">
+<meta data-hr="true" property="og:description" content="$rmetas[og_description]">
+<meta data-hr="true" property="og:image" content="$rmetas[og_image]">
 
-if ($url == "" or !preg_match('/^http/',$url) ) {
-  echo "</head><body>\nError - bad url: '<b>$url</b>'<br>\nuse <b>". $_SERVER['REQUEST_SCHEME'] .'://'. $_SERVER['HTTP_HOST'] . explode('?', $_SERVER['REQUEST_URI'], 2)[0] . "?r=URL</b> please with http or https in url\n\n</body></html>\n";
```

Comment:
Seems like a good thing to have the URL used for the script to report back to how it should be used on failure.

Reply:
Yeah, good point, I'll add back an `echo` with the URL but unless someone examines the HTTP server logs, it's just gonna be noise.

Reply:
Updated in e59a29b